### DISCUSSION TITLE
error in "Customize Access Token (JWT) Claims hook" from RBAC toturial

### DISCUSSION THREAD

**Discussion by mendlero** [supabase/supabase#28852] [Questions] [answered]
I followed the [RBAC toturial in the docs](https://supabase.com/docs/guides/database/postgres/custom-claims-and-role-based-access-control-rbac?queryGroups=language&language=plpgsql) and created the following function:
```plpgsql
  DECLARE
    claims    jsonb;
    user_role public.app_role;
  BEGIN
    select role into user_role from public.user_roles where user_id = (event->>'user_id')::uuid;

    claims := event->'claims';

    if user_role is not null then
      claims := jsonb_set(claims, '{user_role}', to_jsonb(user_role));
    else
      claims := jsonb_set(claims, '{user_role}', 'null');
    end if;

    event := jsonb_set(event, '{claims}', claims);

    return event;
  END;
```
just as described in the docs, but I've been getting an error every single time I try to sign in a user:
`AuthApiError: Error running hook URI: pg-functions://postgres/public/custom_access_token_hook`

Does anyone know what the problem is?

IDK if that matters but I changed the user_roles table to look like this (It fits my needs better):
```sql
CREATE TABLE if not exists user_roles (
  user_id     UUID primary key references auth.users ON DELETE CASCADE,
  role        app_role NOT NULL
);
```

**Comment by GaryAustin1** [comment]
Are you getting an error in the dashboard Postgres or Auth logs with details?

**Comment by mendlero** [comment]
Now that I looked at the logs, I found that the error was a permission denied for table user_roles.
Which role do I need to give permissions to access the table?
Also, the security of the function is `secority invoker`, should it be like that?

**Reply by GaryAustin1** [reply, marked as answer]
Seems like you are missing one or more of these: 
![image](https://github.com/user-attachments/assets/520d0aee-1356-4bd8-b826-efbe37eb545c)

Security Definer type would have also solved the grants to supabase_auth_admin.   I don't know why they chose to do grants versus security definer, although normally safer if not a trigger function.
